internal/worker/controlsocket/worker.go

```diff
@@ -68,58 +85,252 @@ func NewWorker(config Config) (worker.Worker, error) {
+	if errors.Is(err, io.EOF) {
+		w.writeResponse(resp, http.StatusBadRequest, errorf("missing request body"))
+		return
+	} else if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, errorf("request body is not valid JSON: %v", err))
+		return
+	}
+
+	code, err := w.addMetricsUser(parsedBody.Username, parsedBody.Password)
+	if err != nil {
+		w.writeResponse(resp, code, errorf("%v", err))
+		return
+	}
 
-	srv := http.Server{Handler: router}
-	defer func() {
-		err := srv.Close()
+	w.writeResponse(resp, code, infof("created user %q", parsedBody.Username))
+}
+
+func (w *Worker) addMetricsUser(username, password string) (int, error) {
+	err := validateMetricsUsername(username)
+	if err != nil {
+		return http.StatusBadRequest, err
+	}
+
+	if password == "" {
+		return http.StatusBadRequest, errors.NotValidf("empty password")
+	}
+
+	user, err := w.config.State.AddUser(username, username, password, userCreator)
+	cleanup := true
+	// Error handling here is a bit subtle.
+	switch {
+	case errors.Is(err, errors.AlreadyExists):
+		// Retrieve existing user
+		user, err = w.config.State.User(names.NewUserTag(username))
 		if err != nil {
-			w.config.Logger.Warningf("error closing HTTP server: %v", err)
+			return http.StatusInternalServerError,
+				fmt.Errorf("retrieving existing user %q: %v", username, err)
 		}
-	}()
-
-	go func() {
-		// Wait for the tomb to start dying and then shut the server down.
-		<-w.tomb.Dying()
-		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
-		defer cancel()
-		if err := srv.Shutdown(ctx); err != nil {
-			w.config.Logger.Warningf("error shutting down HTTP server: %v", err)
+
+		// We want this operation to be idempotent, but at the same time, this
+		// worker shouldn't mess with users that have not been created by it.
+		// So ensure the user is identical to what we would have created, and
+		// otherwise error.
+		if user.CreatedBy() != userCreator {
+			return http.StatusConflict, errors.AlreadyExistsf("user %q (created by %q)", user.Name(), user.CreatedBy())
 		}
-	}()
+		if !user.PasswordValid(password) {
+			return http.StatusConflict, errors.AlreadyExistsf("user %q", user.Name())
+		}
+
+	case err == nil:
+		// At this point, the operation is in a partially completed state - we've
+		// added the user, but haven't granted them the correct model permissions.
+		// If there is an error granting permissions, we should attempt to "rollback"
+		// and remove the user again.
+		defer func() {
+			if cleanup == false {
+				// Operation successful - nothing to clean up
+				return
+			}
+
+			err := w.config.State.RemoveUser(user.UserTag())
+			if err != nil {
+				// Best we can do here is log an error.
+				w.config.Logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+					username, err)
+			}
+		}()
+
+	default:
+		return http.StatusInternalServerError, errors.Annotatef(err, "failed to create user %q: %v", username, err)
```

Comment:
Annotation is what you were doing, as is the case for the others you've got proximate to here. The fact that it failed is implicit in the error, so: "creating user...".

Reply:
Changed, though this was code moved from `handler.go`, I didn't write it.